### PR TITLE
Translate scala unisave

### PR DESCRIPTION
**Pull request by eddturner**
repo based entirely on original unisave's orm, and uses same code, except it is slimmed down

I can talk you through this to begin with if it helps, because it is something completely new (but relatively simple in terms of logic)

### CODE REVIEW

File: unisave-rest/src/main/java/org/uniprot/api/unisave/repository/domain/impl/ReleaseImpl.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            query = "SELECT r from Release r where r.releaseNumber=:rel"),
+    @NamedQuery(
+            name = "ReleaseImpl.findAllRelease",
+            query = "SELECT r from Release r where r.id <> 9999999999 order by r.id desc")
+})
+@Data
+public class ReleaseImpl implements Release {
+
+    public enum Query {
+        findReleaseByReleaseNumber,
```
Comment (LeonardoGonzales) [inline unisave-rest/src/main/java/org/uniprot/api/unisave/repository/domain/impl/ReleaseImpl.java:27]: sonar will complain about lower case enums, it is not a blocker, I think major only

File: unisave-rest/src/main/java/org/uniprot/api/unisave/request/UniSaveRequest.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Created 27/03/20
+ *
+ * @author Edd
+ */
+public class UniSaveRequest {
+    @Data
+    public static class Entries {
+        private String accession;
```
Comment (LeonardoGonzales) [inline unisave-rest/src/main/java/org/uniprot/api/unisave/request/UniSaveRequest.java:16]: not sure if necessary here, but maybe validate the accession input with the famous Regex...

File: unisave-rest/src/main/java/org/uniprot/api/unisave/service/impl/UniSaveServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+    private static final int CURRENT_DATE_UPDATE_FREQUENCY = 3600 * 12;
+    private final UniSaveRepository repository;
+    private String currentReleaseDate;
+
+    @Autowired
+    UniSaveServiceImpl(UniSaveRepository repository) {
+        this.repository = repository;
+    }
+
+    //    public void initCurrentReleaseDate() {
```
Comment (LeonardoGonzales) [inline unisave-rest/src/main/java/org/uniprot/api/unisave/service/impl/UniSaveServiceImpl.java:46]: can we remove commented code?

File: unisave-rest/pom.xml [hunk 4]
```diff
@@ -0,0 +1,92 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<parent>
+		<artifactId>uniprot-rest-api</artifactId>
+		<groupId>org.uniprot</groupId>
+		<version>1.0.0-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (eddturner) [reply unisave-rest/pom.xml:8]: thank u!

File: unisave-rest/src/main/java/org/uniprot/api/unisave/controller/UniSaveController.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    // responses in fasta/flatfile format will require fetching
+    // content regardless of what user specifies, so set it
+    private void setContentIfRequired(UniSaveRequest.Entries uniSaveRequest, String acceptHeader) {
+        if (acceptHeader.equals(FASTA_MEDIA_TYPE_VALUE)
+                || acceptHeader.equals(FF_MEDIA_TYPE_VALUE)) {
+            uniSaveRequest.setIncludeContent(true);
+        }
+    }
+
+    private String getHeader(HttpServletRequest servletRequest) {
```
[the remark this answers is not in the document]
Comment (eddturner) [reply unisave-rest/src/main/java/org/uniprot/api/unisave/controller/UniSaveController.java:112]: done

File: unisave-rest/src/main/java/org/uniprot/api/unisave/repository/UniSaveRepository.java [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+                throw new ResourceNotFoundException("No entries for " + accession + " were found");
+            }
+
+            for (EntryImpl entryImpl : resultList) {
+                setContent(entryImpl, session);
+            }
+
+            return resultList;
+        } catch (ResourceNotFoundException e) {
+            throw e;
```
[the remark this answers is not in the document]
Comment (eddturner) [reply unisave-rest/src/main/java/org/uniprot/api/unisave/repository/UniSaveRepository.java:62]: good point. i've now followed same pattern i did in other repository methods, where the error is logged first.

File: unisave-rest/src/main/java/org/uniprot/api/unisave/repository/UniSaveRepository.java [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+                if (status.getEventType() == EventTypeEnum.deleted) deleted.add(status);
+                else if (status.getEventType() == EventTypeEnum.merged) merged.add(status);
+                else if (status.getEventType() == EventTypeEnum.replacing) replacing.add(status);
+            }
+
+            // find the replacing accession
+            for (Object entryQueryResult : entryQueryResultList) {
+                EntryInfoImpl entryInfo = convertFromObjectArray((Object[]) entryQueryResult);
+
+                if (!replacing.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (eddturner) [reply unisave-rest/src/main/java/org/uniprot/api/unisave/repository/UniSaveRepository.java:128]: done

File: unisave-rest/src/main/java/org/uniprot/api/unisave/service/impl/UniSaveServiceImpl.java (hunk 3 above)
Comment (eddturner) [reply unisave-rest/src/main/java/org/uniprot/api/unisave/service/impl/UniSaveServiceImpl.java:46]: done

File: unisave-rest/src/main/java/org/uniprot/api/unisave/request/UniSaveRequest.java (hunk 2 above)
Comment (eddturner) [reply unisave-rest/src/main/java/org/uniprot/api/unisave/request/UniSaveRequest.java:16]: done (for all uses of an accession in endpoints)

File: unisave-rest/src/main/java/org/uniprot/api/unisave/repository/domain/impl/ReleaseImpl.java (hunk 1 above)
Comment (eddturner) [reply unisave-rest/src/main/java/org/uniprot/api/unisave/repository/domain/impl/ReleaseImpl.java]: done

File: unisave-rest/src/main/java/org/uniprot/api/unisave/repository/domain/impl/EntryImpl.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            query =
+                    "SELECT e.database, e.accession, e.name, e.entryVersion, e.entryMD5, e.sequenceVersion, e.sequenceMD5, "
+                            + "r1, r2 from Entry e left join e.firstRelease r1 left join e.lastRelease r2 WHERE e.accession=:acc and e.database <> org.uniprot.api.unisave.repository.domain.DatabaseEnum.WrongTrembl"
+                            + " ORDER BY e.entryVersion DESC"),
+})
+@Data
+public class EntryImpl implements Entry {
+
+    public enum Query {
+        findEntryByAccessionAndEntryId,
```
[the remark this answers is not in the document]
Comment (eddturner) [reply unisave-rest/src/main/java/org/uniprot/api/unisave/repository/domain/impl/EntryImpl.java]: done

File: unisave-rest/src/main/java/org/uniprot/api/unisave/repository/domain/EventTypeEnum.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package org.uniprot.api.unisave.repository.domain;
+
+/**
+ * Created with IntelliJ IDEA. User: wudong Date: 12/11/2013 Time: 14:20 To change this template use
+ * File | Settings | File Templates.
+ */
+public enum EventTypeEnum {
```
[the remark this answers is not in the document]
Comment (eddturner) [reply unisave-rest/src/main/java/org/uniprot/api/unisave/repository/domain/EventTypeEnum.java:7]: yep thanks, i'd just had whatever wudong did, and didn't really think about it. thank you! done now.

File: unisave-rest/src/main/java/org/uniprot/api/unisave/repository/domain/ContentTypeEnum.java [hunk 10]
```diff
@@ -0,0 +1,6 @@
+package org.uniprot.api.unisave.repository.domain;
+
+public enum ContentTypeEnum {
+    Full,
```
[the remark this answers is not in the document]
Comment (eddturner) [reply unisave-rest/src/main/java/org/uniprot/api/unisave/repository/domain/ContentTypeEnum.java]: done

File: unisave-rest/src/main/java/org/uniprot/api/unisave/repository/domain/DatabaseEnum.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
+package org.uniprot.api.unisave.repository.domain;
+
+public enum DatabaseEnum {
+    Swissprot {
```
[the remark this answers is not in the document]
Comment (eddturner) [reply unisave-rest/src/main/java/org/uniprot/api/unisave/repository/domain/DatabaseEnum.java]: done

**Pull request merged**